resources/components/base/Button.vue

```diff
@@ -0,0 +1,242 @@
+<template>
+	<button
+		class="wbmad-button"
+		v-bind:class="classObject"
+		v-bind:disabled="disabled"
+		v-bind:title="title"
+		v-on:click="$emit( 'click' )"
+	>
+		<icon
+			v-if="icon"
+			v-bind:icon="icon"
+			v-bind:invert="invert"
+		/>
+		<slot />
+	</button>
+</template>
+
+<script>
+var Icon = require( './Icon.vue' );
+
+module.exports = {
+	name: 'Button',
+
+	components: {
+		icon: Icon
+	},
+
+	props: {
+		classes: {
```

Comment:
I don't think we need a `classes` prop at all here. OOUI needs this because the button elements are created imperatively, and there isn't a template to add classes to directly.

With Vue, we can apply any classes that are needed in the template where this component is consumed, so there is no need to pass class props. For example, this works fine:
```
<my-btn class="publish-button">Publish</my-btn>
```

If the top-level element inside `<my-btn>'s` `<template>` was `button`, and a "button" class was already applied there, the final output would be:
```
<div class="button publish-button">Publish</button>
```

I'd say that pre-defined options for appearance/behavior (framed, disabled, primary, etc) are good candidates for props treatment. That saves the user from having to worry about which CSS classes are actually applied. But for other arbitrary classes there is no need to handle them as props.

Reply:
Oh I was way over-complicating this, thanks for the explanation!